core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -995,21 +1003,10 @@ private[spark] class BlockManager(
       data: () => Either[Array[Any], ByteBuffer]): Unit = {
 
     logInfo(s"Dropping block $blockId from memory")
-    val info = blockInfo.get(blockId)
-
-    // If the block has not already been dropped
-    if (info != null) {
-      info.synchronized {
-        // required ? As of now, this will be invoked only for blocks which are ready
-        // But in case this changes in future, adding for consistency sake.
-        if (!info.waitForReady()) {
-          // If we get here, the block write failed.
-          logWarning(s"Block $blockId was marked as failure. Nothing to drop")
-          return
-        } else if (blockInfo.asScala.get(blockId).isEmpty) {
-          logWarning(s"Block $blockId was already dropped.")
-          return
-        }
+    blockInfoManager.lockForWriting(blockId) match {
```

Comment:
I haven't forgotten about this comment; this is going to be addressed as part of cleanup to move the lock acquisition out of this method so that it takes place higher up the call chain.